optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -61,58 +62,48 @@
-                        sumLong((machineCapacity, processAssignment) ->
-                                processAssignment.getUsage(machineCapacity.getResource()))
-                )
-                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
+                        equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine))
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity.getMachine(),
+                        (machineCapacity, processAssignment) -> machineCapacity,
+                        sumLong((machineCapacity, processAssignment) -> processAssignment
+                                .getUsage(machineCapacity.getResource())))
+                .filter(((machine, machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
                 .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
                         HardSoftLongScore.ONE_HARD,
-                        (machineCapacity, usage) -> usage - machineCapacity.getMaximumCapacity());
+                        (machine, machineCapacity, usage) -> usage - machineCapacity.getMaximumCapacity());
     }
 
-    /**
-     * Conflict: Processes of the same service must run on distinct machines.
-     */
-    private Constraint serviceConflict(ConstraintFactory factory) {
-        return factory.fromUniquePair(MrProcessAssignment.class,
-                equal(MrProcessAssignment::getMachine),
-                equal(MrProcessAssignment::getService)
-        ).penalize(MrConstraints.SERVICE_CONFLICT,
-                HardSoftLongScore.ONE_HARD);
+    protected Constraint serviceConflict(ConstraintFactory factory) {
+        return factory.fromUnfiltered(MrProcessAssignment.class)
+                .join(factory.fromUnfiltered(MrProcessAssignment.class),
+                        equal(MrProcessAssignment::getMachine, MrProcessAssignment::getMachine))
+                .filter((left, right) -> left.getService().equals(right.getService()) && right.getId() > left.getId())
+                .penalize(MrConstraints.SERVICE_CONFLICT, HardSoftLongScore.ONE_HARD);
     }
 
     /**
      * Spread: Processes of the same service must be serviceLocationSpread out across locations.
      */
-    private Constraint serviceLocationSpread(ConstraintFactory factory) {
-        return factory.from(MrProcessAssignment.class)
-                .groupBy(MrProcessAssignment::getService,
-                        ConstraintCollectors.countDistinct(MrProcessAssignment::getLocation))
-                .filter((service, distinctLocationCount) -> service.getLocationSpread() > distinctLocationCount)
-                .penalize(MrConstraints.SERVICE_LOCATION_SPREAD,
-                        HardSoftLongScore.ONE_HARD);
+    protected Constraint serviceLocationSpread(ConstraintFactory factory) {
+        throw new UnsupportedOperationException("ConstraintCollectors.countDistinct should not count null as a distinct value");
```

Comment:
Which MrProcessAssignment have a null location? The `from` already filters out all those that doesn't have a machine. Are there machines with a null location? If so, just add a filter before the groupBy.

Reply:
See the discussion on Zulip chat - this is about uninitialized solution and the fact that a DRL score function works differently in such a condition.